### PR TITLE
Enhancement: bulk edit object permissions

### PR DESCRIPTION
**Pull request by shamoon**
## Proposed change

This PR adds support for bulk updating permissions on a set of objects (Tags, Doc Types, etc.). It uses a lot of existing stuff so should be relatively solid.

Screenshot 2023-09-13 at 1 03 53 AM
Screenshot 2023-09-13 at 1 04 31 AM
Screenshot 2023-09-13 at 1 04 44 AM

Fixes #3280

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain):

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by shamoon** [comment]
Weird, mocking that failure works fine locally...
https://github.com/paperless-ngx/paperless-ngx/actions/runs/6169938973/job/16745227099?pr=4176#step:10:52

**Comment by stumpylog** [comment]
Noticed a couple things

- When I bulk select something, make a change, then hit permissions again, that pop-up doesn't reflect the changes just made.  In the edit for an individual item, the new permissions are reflected.  The API response after bulk edit does look to include the info
- This might be since I don't really use permissions, but when i grant testuser permissions on a tag, I don't see a Tags link when logged in as them.  Even setting them as owner from a super user, the sidebar remains empty.  This is with bulk edit or single edit

**Comment by shamoon** [comment]
- Yea bulk editing permissions is like completely replacing them. It doesn’t take the actual objects current permissions into account. So I think that’s normal (assuming I understood you correctly)
- For the second note, does testuser have global permissions for tags? If not it won’t show in sidebar (global don’t determine object-level). Again, if I understood correctly

File: src/documents/views.py [hunk 1]
```diff
@@ -1205,3 +1207,44 @@ def serve_file(doc: Document, use_archive: bool, disposition: str):
...
+
+        if not user.is_superuser:
+            objs = object_class.objects.filter(pk__in=object_ids)
+            has_perms = all((obj.owner == user or obj.owner is None) for obj in objs)
+
+            if not has_perms:
+                return HttpResponseForbidden("Insufficient permissions")
+
+        try:
+            qs = Tag.objects.filter(id__in=object_ids)
```
[the remark this answers is not in the document]
Comment (shamoon) [reply src/documents/views.py]: Great catch thanks! Tests missed this because they only checked changing tag owner, updated that too

**Comment by shamoon** [comment]
Sorry was just cleaning up the commits, thanks for the catch

**Review by stumpylog — APPROVED** [review]
Looks good, other stuff was just me not using permissions much.  After the user was updated, things changed as expected

**Pull request merged**